app/src/context/slices/projectSlice.ts

```diff
@@ -0,0 +1,1986 @@
+        );
+      }
+      if (!project.database.remote) {
+        throw new Error(
+          `Failed to deactivate active project due to missing remote database. Server ID: ${payload.serverId}. Project ID: ${payload.projectId}`
+        );
+      }
+
+      // close and remove sync (if it's available)
+      const syncId = project.database.remote.syncId;
+      if (syncId) {
+        databaseService.closeAndRemoveSync(syncId);
+      }
+
+      // establish ID of remote DB
+      const remoteDatabaseId = project.database.remote.remoteDbId;
+      // remove (no need to wipe/clean records)
+      databaseService.closeAndRemoveRemoteDatabase(remoteDatabaseId);
+
+      // establish ID of local DB
+      const localDatabaseId = project.database.localDbId;
+      // wipe and remove local database (cleaning records)
+      databaseService.closeAndRemoveLocalDatabase(localDatabaseId, {
+        clean: true,
+      });
+
+      // updates the state with all of this new information
+      state.servers[payload.serverId].projects[payload.projectId] = {
+        // These are retained
+        metadata: project.metadata,
+        projectId: project.projectId,
+        rawUiSpecification: project.rawUiSpecification,
+        uiSpecificationId: project.uiSpecificationId,
+        createdAt: project.createdAt,
+        lastUpdated: project.lastUpdated,
+        serverId: project.serverId,
+
+        // These are updated (to indicate de-activation)
+        isActivated: false,
+        database: undefined,
+      };
+    },
+
+    /**
+     * Updates the database auth for a given project.
+     *
+     * This involves
+     *
+     * - destroy and deregister sync
+     * - destroy and deregister remote DB
+     * - create remote DB with new connection config
+     * - create sync with new remote DB (and existing local DB)
+     * - register new entries
+     *
+     * This method is run when the token is refreshed.
+     */
+    updateDatabaseAuth: (
+      state,
+      action: PayloadAction<{jwtToken: string} & ProjectIdentity>
+    ) => {
+      // check project/server exists
+      const payload = action.payload;
+
+      // Check the server exists
+      const server = serverById(state, payload.serverId);
+      if (!server) {
+        // abort
+        throw new Error(
+          `You cannot update connection for a project for a server which does not exist. Server ID: ${payload.serverId}. Project ID: ${payload.projectId}`
+        );
+      }
+
+      // Check the couch DB url has been populated
+      if (!server.couchDbUrl) {
+        // abort
+        throw new Error(
+          `Cannot update connection when we don't know the couchDBUrl. Server ID: ${payload.serverId}. Project ID: ${payload.projectId}`
+        );
+      }
+
+      // check the project exists
+      const project = projectByIdentity(state, payload);
+      if (!project) {
+        // abort
+        throw new Error(
```

Comment:
After reading this for the nth time I wonder if you can't factor all of the checks out to a helper fn?

Reply:
I don't think you can because they also behave as type assertions to the type checker. If you put into a separate function typescript is not smart enough to see that.